components/script/dom/readablebytestreamcontroller.rs

```diff
@@ -2,19 +2,343 @@
+use crate::js::rust::CustomTrace;
+use crate::realms::{enter_realm, InRealm};
+use crate::script_runtime::{JSContext, JSContext as SafeJSContext};
+
+/// <https://streams.spec.whatwg.org/#readable-byte-stream-queue-entry>
+#[derive(JSTraceable)]
+pub struct ReadableByteStreamQueueEntry {
+    buffer: ArrayBuffer,
+    byte_offset: usize,
+    byte_length: usize,
+}
+
+/// <https://streams.spec.whatwg.org/#queue-with-sizes>
+#[derive(Default, JSTraceable, MallocSizeOf)]
+pub struct QueueWithSizes {
+    total_size: usize,
+    #[ignore_malloc_size_of = "ReadableByteStreamQueueEntry"]
+    queue: VecDeque<ReadableByteStreamQueueEntry>,
+}
+
+impl QueueWithSizes {
+    /// <https://streams.spec.whatwg.org/#dequeue-value>
+    fn dequeue_value(&mut self) -> ReadableByteStreamQueueEntry {
+        self.queue
+            .pop_front()
+            .expect("Buffer cannot be empty when dequeue value is called into.")
+    }
+
+    /// <https://streams.spec.whatwg.org/#enqueue-value-with-size>
+    fn enqueue_value_with_size(&mut self, value: ReadableByteStreamQueueEntry) {
+        self.queue.push_back(value);
+    }
+
+    fn is_empty(&self) -> bool {
+        self.queue.is_empty()
+    }
+}
+
+/// The fulfillment handler for
+/// <https://streams.spec.whatwg.org/#readable-byte-stream-controller-call-pull-if-needed>
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[allow(crown::unrooted_must_root)]
+struct PullAlgorithmFulfillmentHandler {
+    // TODO: check the validity of using Dom here.
+    controller: Dom<ReadableByteStreamController>,
+}
+
+impl Callback for PullAlgorithmFulfillmentHandler {
+    /// Handle fufillment of pull algo promise.
+    fn callback(&self, _cx: JSContext, _v: HandleValue, _realm: InRealm) {
+        todo!();
+    }
+}
+
+/// The rejection handler for
+/// <https://streams.spec.whatwg.org/#readable-byte-stream-controller-call-pull-if-needed>
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[allow(crown::unrooted_must_root)]
+struct PullAlgorithmRejectionHandler {
+    // TODO: check the validity of using Dom here.
+    controller: Dom<ReadableByteStreamController>,
+}
+
+impl Callback for PullAlgorithmRejectionHandler {
+    /// Handle rejection of pull algo promise.
+    fn callback(&self, _cx: JSContext, _v: HandleValue, _realm: InRealm) {
+        todo!();
+    }
+}
+
+// FIXME: this is similar to [`script::dom::readablestream::ReaderType`] but without reader
+#[derive(JSTraceable)]
+enum ReaderType {
+    BYOB,
+    Default,
+}
+
+// TODO: temporary implementation, should change after having real usage
+#[derive(JSTraceable)]
+enum ViewConstructor {
+    TypedArray,
+    DataView,
+}
+
+#[derive(JSTraceable, MallocSizeOf)]
+struct PullIntoDescriptor {
+    buffer: Vec<u8>,
+    buffer_byte_length: usize,
+    byte_offset: usize,
+    byte_length: usize,
+    bytes_filled: usize,
+    minimun_fill: usize,
+    element_size: usize,
+    #[ignore_malloc_size_of = "ViewConstructor"]
```

Comment:
I'm adding `ignore_malloc_size_of` just because the compiler is complaining about it, but I don't know when to use it. Am I misusing it?

Reply:
You can try to derive it on the enum, but again, look at [`dom/bindings/buffer_source.rs`](https://github.com/servo/servo/blob/main/components/script/dom/bindings/buffer_source.rs).